Change the return type and semantics of the `create_{accounts,transfers}`.

batiati opened the pull request:
### Motivation:
The previous `create_{accounts,transfers}` operations had sparse returns, including only results for errors and omitting successfully created events. While this was a good design choice to reduce the number of bytes transmitted by the wire protocol, applications often need to know _when_ an event was committed (or validated and failed) by the TigerBeetle cluster.  
Without returning the timestamp, this required an additional network trip to look up the just-created event by ID.

### This PR:
This PR updates the `CreateAccountsResult` and `CreateTransfersResult` types to include the `timestamp` of the processed event. It also removes the `index` field, since each event now has a corresponding result at its event index.

- For successfully created events (`.ok`), the returned timestamp is the cluster-assigned value. In the case of imported events, it is the user-defined timestamp.  
- For existing events (`.exists`), the returned timestamp matches that of the original object, allowing applications to handle `.exists` as `.ok`.  
- For failed events, the returned timestamp indicates _when_ validation occurred, enabling users to perform point-in-time queries, for example, to determine an account’s balance at that moment or which transfers had occurred up to that point.

### Breaking changes:

This introduces a series of API breaking changes that affect all client libraries.
The previous behavior is still supported for compatibility with **old clients**.

> TODO: Update the tracking issue 2231 for API changes.

### Additional API changes:

#### NodeJS

- Renamed the result types `CreateAccountError`, `CreateAccountsError`, `CreateTransferError`, `CreateTransfersError`.

#### Java

- Added the annotation `throws InterruptedException` for all blocking methods (See https://github.com/tigerbeetle/tigerbeetle/pull/2650).

#### Dotnet

- Removed non-batched methods `CreateAccount`, `CreateTransfer`, `LookupAccount`, and `LookupTransfer`. 

#### Go

- Renamed the result types `AccountEventResult` and `TransferEventResult`.
- Packages `types` and `errors` moved to the root package.

<details>
<summary>Thoughts (resolved)</summary>

#### 1. Naming things

"CreateAccountResult" is the enum, and "CreateAccount**s**Result" is the struct, but this naming convention can be confusing:

- Only the letter “s” distinguishes the two, even though `CreateAccountsResult` does not need to be plural, as it is already returned as an array.  
- The repetitive pattern `result.result == .ok` emerges when checking the result within the result.

Suggestions:

- Rename the enum to `CreateAccountResultCode` or `CreateAccountStatus`, so we can write `result.code == .ok` (though note that `code` might be overloaded with `account.code`) or `result.status == .ok`.  
- Rename the struct to `CreateAccountResponse`, allowing `response.result == .ok`.
  Additionally, “Response” conveys that the type includes more information, such as the `timestamp`, rather than just the outcome.

#### 2. Compatibility

The cluster remains compatible with clients as old as version `0.16.4` for as long as necessary, giving users sufficient time to upgrade their applications.

However, we could also make the client libraries support both operations simultaneously within the same version, giving users the option to migrate incrementally, and benefit from client bugfixes without having to upgrade to the new API.

I believe it's not good for the following reasons:
- It increases the surface area (size/tests/documentation) of the client libraries.  
- The two versions of the same operation would not be batched together, discouraging partial migrations.  
- To avoid introducing breaking changes now, we would end up deferring the breaking changes to a future release anyway, so why not now?

#### 3. `getHeader()` API

- Would the new API replace https://github.com/tigerbeetle/tigerbeetle/pull/2481?

</details>

matklad commented on src/state_machine.zig:
```diff
@@ -2637,7 +2684,7 @@ pub fn StateMachineType(
             const result_max: u32 = @max(
                 operation_event_max(.create_transfers, self.batch_size_limit),
-                operation_event_max(.deprecated_create_transfers, self.batch_size_limit),
+                operation_event_max(.deprecated_create_transfers_unbatched, self.batch_size_limit),
```
Might be worth it to file rename as a separate PR, just to reduce the diff for easier review.

matklad commented on src/state_machine.zig:
```diff
@@ -4197,7 +4319,7 @@ pub fn StateMachineType(
 
         fn post_or_void_pending_transfer_exists(
-            t: *const Transfer,
+            t: *Transfer,
```
These `*const T` -> `*T` changes make the blade glow blue, mutability is hard. Is this so that we can assign timestamp? 

Maybe we should return `struct { CreateTransferResult, u64 }` from this function, and assign timestamp at the call-site, to make it clear that the timestamp is the only part that can change? Ah, we could return `CreateAccountsResult` directly, no?

batiati replied:
I found there an interesting opportunity to remove one copy.

https://github.com/tigerbeetle/tigerbeetle/blob/c47d8c9778c5e9bb39c0e87114624269d2b880c5/src/state_machine.zig#L3845-L3847

Actually, we copy it twice, also from the outer `for` loop to assign the timestamp.

batiati committed:
[StateMachine] Return the timestamp.
Change the return type and semantics of the `create_accounts` and `create_transfers` operations.

The previous `create_*` operations returned sparse results containing only errors, omitting success results.
While this is a good design choice to reduce the number of bytes transmitted by the wire protocol, in many cases the application needs to know _when_ the event was committed by the TigerBeetle cluster. Without returning the timestamp, this would require another network trip to look up the just-created event by ID.

This commit changes the return type to include the `timestamp` of the processed event. It also removes the `index` field, since now each event, both `ok` and errors, must have a corresponding result at the event’s index.

For successfully created events, the returned timestamp is the cluster timestamp used as a unique identifier.
For imported events, it is the user-defined timestamp.
For `exists` events,  it returns the timestamp of the original event.
For other failed events, the returned timestamp reflects _when_ the validation happened, allowing users to run point-in-time queries to diagnose, for example, what the balance was at that time or which transfers had taken place so far.

The state machine still supports the previous behavior for old clients.

batiati committed:
[Repl] Support `create_*_with_results`

batiati committed:
[StateMachine] Fuzzers, unit tests and VOPR support for `create_*_with_results`

batiati committed:
[tb_client] Support `create_*_with_results`

batiati committed:
[Java] Support `create_*_with_results`

batiati committed:
[NodeJS] Support `create_*_with_results`
Additional API changes, renaming the following types:
CreateAccountError => CreateAccountResult
CreateAccountsError => CreateAccountsResult
CreateTransferError => CreateTransferResult
CreateTransfersError => CreateTransfersResult

batiati committed:
[Go] Support `create_*_with_results`
Additional API changes, renaming the following types:
AccountEventResult => CreateAccountsResult
TransferEventResult => CreateTransfersResult

batiati committed:
[Dotnet] Support `create_*_with_results`
Additional API changes, remove the non-batched methods `CreateAccount`, `CreateTransfer`, `LookupAccount` and `LookupTransfer`.

batiati committed:
[Python] Support `create_*_with_results`

batiati committed:
[Rust] Support `create_*_with_results`

batiati committed:
[Vortex] Support `create_*_with_results`

batiati committed:
[Documentation] Explain the new return type for `create_*_with_results`.

batiati committed:
[AOF] Support `create_*_with_results`
Also includes other deprecated operations that need to be replayed.

batiati committed:
[Java] Throws InterruptedException
Not related to `create_*_with_results`, but an API breaking change that we need to introduce.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -3729,13 +3909,13 @@ pub fn StateMachineType(
-                .timestamp = timestamp,
+                .timestamp = timestamp_actual,
             });
-            self.commit_timestamp = timestamp;
-            return .ok;
+            self.commit_timestamp = timestamp_actual;
```
Ahh ok, so previously `timestamp_event` used the import timestamp, but now it is always the position-based timestamp, so here we need to override it.

sentientwaffle commented on src/state_machine/workload.zig:
```diff
@@ -1465,6 +1501,72 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                }
+            }
+
+            for (transfers, results) |*transfer, *outcome| {
```
We can assert that the `outcome.reserved == 0`. Likewise for `create_accounts`.

sentientwaffle commented on src/clients/c/samples/main.c:
```diff
@@ -163,11 +161,11 @@ int main(int argc, char **argv) {
-        packet.status = TB_PACKET_OK;                      // Will be set when the reply arrives.
+        packet.operation = TB_OPERATION_CREATE_TRANSFERS_WITH_RESULTS; // The operation to execute.
+        packet.data = transfers;                                       // The data to be sent.
+        packet.data_size = TRANSFERS_SIZE;                             //
```
Nice catch!

sentientwaffle commented on src/testing/vortex/workload.zig:
```diff
@@ -161,27 +161,18 @@ fn reconcile(result: Result, command: *const Command, model: *Model) !void {
     switch (result) {
         .create_accounts => |entries| {
```
Maybe s/entries/results? Or (`account_results`/`transfer_results`)

sentientwaffle commented on docs/reference/requests/create_accounts.md:
Re: Compatibility:

I agree that clients shouldn't support both simultaneously. People can just keep using the old client until they are ready to switch.

Re: `getHeader()` API:

I was originally thinking yes, but now I think it isn't enough, since it doesn't help with queries/lookups.

sentientwaffle commented on docs/reference/requests/create_accounts.md:
```diff
@@ -4,7 +4,31 @@ Create one or more [`Account`](../account.md)s.
+
+</details>
 
 ## Result
```
Should this header be changed to

```
### `result`
```
to match the timestamp header? And then the codes themselves would transition from `###` to  `####`.

sentientwaffle reviewed (commented):
Note that there is a conflict with main.

batiati committed:
Rename timestamps for clarity
- `event_timestamp` to `timestamp_event`
- `timestamp` to `timestamp_event` when it's relative to the event index.
- `expiry` to `timestamp_expiry`

batiati committed:
[Docs] Reorganize `result` and enums

batiati committed:
Review

batiati committed:
Drop the temporary `_with_results` suffix.
For refactoring, it was used to distinguish the operation name from other identifiers containing the `create_transfers`/`create_accounts` string.

batiati committed:
Rename: `result` to `status` and `.ok` to `.created`.
- `Create{Account,Transfer}Result` to `Create{Account,Transfer}Status`.
- `Create{Accounts,Transfers}Result` to `Create{Account,Transfer}Result`, and rename the field `result` to `status`.
- Status code `.ok` to `.created`.

batiati committed:
[Docs] Rename: `result` to `status` and `.ok` to `.created`.

batiati commented:
More API changes:
- Renamed the enum to `Create{Account,Transfer}Status`, and the field `result` to `status`.
- Also renamed the result `ok` to `created`.
- Dropped the Go client packages `types` and `errors`, declaring all types together.

Not an API change, but also included:
- Assigned a non-zero value to the result `created` as we are now materializing it to the wire protocol.

batiati committed:
[Go] Move errors and types to the root package and remove the `pkg` subfolder.

batiati committed:
[Docs] Simplify handling errors in Go.

batiati committed:
Assign a non-zero value to the `created` status.

sentientwaffle commented on docs/coding/requests.md:
```diff
@@ -35,7 +35,7 @@ Each request has a corresponding _event_ and _result_ type:
 | ----------------------- | --------------------------------------------------------------------- | ----------------------------------------------------------------------------------- |
 | `create_accounts`       | [`Account`](../reference/requests/create_accounts.md#event)           | [`CreateAccountResult`](../reference/requests/create_accounts.md#result)            |
-| `create_transfers`      | [`Transfer`](../reference/requests/create_transfers.md#event)         | [`CreateTransferResult`](../reference/requests/create_transfers.md#result)          |
+| `create_transfers`      | [`Transfer`](../reference/requests/create_transfers.md#event)         | [`CreateAccountResult`](../reference/requests/create_transfers.md#result)           |
```
This should still be `CreateTransferResult`?

sentientwaffle commented on src/clients/go/errors.go:
Now that this isn't an internal module, I think we should make the error types private (e.g. `type errUnexpected`) so that they can't be constructed by dependent packages.

batiati replied:
I think error types must be public, because users might need to do something like `if errors.Is(err, tb.ErrMaximumBatchSizeExceeded)`

sentientwaffle reviewed (commented):
WIP review (second pass)

batiati committed:
[Rust] Rebase

batiati committed:
[Docs] Review

sentientwaffle commented on src/clients/java/src/test/java/com/tigerbeetle/BatchTest.java:
```diff
@@ -956,24 +970,29 @@ private static final class DummyTransferDto {
-        dummyCreateAccountResultsStream.putInt(1).putInt(CreateAccountResult.Exists.value); // Item
-                                                                                            // 1 -
-                                                                                            // Exists
+        dummyCreateAccountStatussStream = ByteBuffer.allocate(32).order(ByteOrder.LITTLE_ENDIAN);
```
s/statuss/statuses

sentientwaffle commented on src/clients/python/python_bindings.zig:
```diff
@@ -298,30 +302,41 @@ fn convert_ctypes_to_python(comptime name: []const u8, comptime Type: type) []co
     comptime type_info: anytype,
-    comptime c_name: []const u8,
+    comptime python_name: []const u8,
+    skip_initialization: bool,
```
Maybe rename to `skip_field_defaults`? Or invert it to `has_default_initialization`? I was initially (heh) confused about what this was doing.

sentientwaffle commented on src/devhub/devhub.js:
These should still be `ok`, not `created`.

sentientwaffle commented on src/testing/vortex/workload.zig:
```diff
@@ -22,7 +22,7 @@
 //! 3. The workload receives the results, and expects them to be of the same operation type as
 //!    originally requested. There might be fewer results than events, because clients can omit
-//!    .ok results.
+//!    .created results.
```
You can remove the last sentence of this comment.

sentientwaffle commented on src/testing/vortex/workload.zig:
```diff
@@ -159,69 +159,48 @@ fn operation_from_command(tag: std.meta.Tag(Command)) StateMachine.Operation {
         },
-        .create_transfers => |entries| {
+        .create_transfers => |transfer_results| {
             const transfers = command.create_transfers;
```
Can we assert that `transfers.len == transfer_results.len`?

sentientwaffle reviewed (commented):
Second review pass done. Other than a few nits this LGTM.

batiati committed:
[Docs] Update the REPL output

batiati committed:
Review

batiati commented:
> Second review pass done. Other than a few nits this LGTM.

Thank you @sentientwaffle, for catching all those hard-to-spot replaces that went wrong!
As it's a major API change, I'll include a migration guide with a general overview/motivation and the specificities of each client library. Then we can decide where this document will live.

batiati commented:
Wait, there's one more API breaking change I wanted to introduce: currently, if the application requests a `limit` that's too large in a query filter, it will silently cap the actual limit to the batch max size. Instead, we should return the error `TooMuchData` (and maybe something similar for `limit = 0`, but not strong about that).


https://github.com/tigerbeetle/tigerbeetle/blob/655602189d6d5ebe268e878aad57cb389c5b4673/src/state_machine.zig#L822-L823

sentientwaffle reviewed (dismissed):
LGTM modulo git conflicts!

batiati commented:
> LGTM modulo git conflicts!

😅🫣😛

Thanks @sentientwaffle! I'll leave the migration guide and the `TooMuchData` error as a follow up PR.

batiati committed:
Merge conflicts

batiati committed:
[Python] Make constants UPPER CASE.
Related https://github.com/tigerbeetle/tigerbeetle/issues/3385

batiati committed:
[Merge] Fix timestamp assertions.

batiati committed:
Merge branch 'main' into batiati/create_results

sentientwaffle commented on src/clients/dotnet/samples/walkthrough/Program.cs:
```diff
@@ -41,8 +40,8 @@
 
-        var accountErrors = client.CreateAccounts(accounts);
-        // Error handling omitted.
+        var accountsResults = client.CreateAccounts(accounts);
```
Should "accounts results" be "account results"? (likewise for transfers).
(Also `accounts_results` in python client.)

batiati replied:
Indeed, it is very inconsistent across the entire codebase, not only examples!

sentientwaffle commented on src/clients/go/ci.zig:
```diff
@@ -39,14 +39,6 @@ pub fn tests(shell: *Shell, gpa: std.mem.Allocator) !void {
     }
 
     try shell.exec("go test", .{});
-    {
-        log.info("testing `types` package helpers", .{});
-
-        try shell.pushd("./pkg/types");
-        defer shell.popd();
-
-        try shell.exec("go test", .{});
-    }
```
🎉

sentientwaffle commented on src/state_machine/auditor.zig:
```diff
@@ -493,174 +493,277 @@ pub const AccountingAuditor = struct {
+            *account,
+            *result,
+            i,
+        | {
```
Why split these lines?

sentientwaffle commented on docs/reference/requests/create_accounts.md:
```diff
@@ -4,44 +4,65 @@ Create one or more [`Account`](../account.md)s.
+Create results are sparse, containing only failed events and the `index`
+of the account within the events batch.
+
+The network protocol does not include an [`created`](#created) result for successfully
```
```suggestion
The network protocol does not include a [`created`](#created) result for successfully
```

sentientwaffle commented on docs/reference/requests/create_transfers.md:
```diff
@@ -6,44 +6,65 @@ amount fields of its [debit](../transfer.md#debit_account_id) and
+Create results are sparse, containing only failed events and the `index`
+of the transfer within the events batch.
+
+The network protocol does not include an [`created`](#created) result for successfully
```
```suggestion
The network protocol does not include a [`created`](#created) result for successfully
```

sentientwaffle commented on src/state_machine_tests.zig:
```diff
@@ -624,6 +635,15 @@ fn check_version(
     var transfers = std.AutoHashMap(u128, Transfer).init(allocator);
     defer transfers.deinit();
 
+    // The result code `.exists` always returns the timestamp of the original event.
+    // Even if the existing event was created within a linked chain and rolled back.
+    // For example, the linked chain:
+    //  events:  { id=1, flags=linked; id=1 }
+    //  results: { result=linked_event_failed, timestamp=100; result=exists, timestamp=100 }
+    //                                                   ^^^                           ^^^
```
🤯 Tricky!

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -2903,63 +2924,93 @@ pub fn StateMachineType(comptime Storage: type) type {
             const batch_imported = events.len > 0 and events[0].flags.imported;
             for (events, 0..) |*event, index| {
-                const result = result: {
+                const result_code, const timestamp_actual = result: {
```
s/code/status, both for consistency  and because "code" means something different in the context of accounts/transfers

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -3774,25 +3921,21 @@ pub fn StateMachineType(comptime Storage: type) type {
             assert(self.forest.grooves.transfers.get(t.id) == .not_found);
             assert(t.flags.padding == 0);
-            assert(timestamp > self.commit_timestamp or t.flags.imported);
-            if (t.flags.imported) {
-                assert(t.timestamp == timestamp);
-            } else {
-                assert(t.timestamp == 0);
-            }
```
Why remove these asserts? Particularly the "!imported implies timestamp=0".

batiati replied:
It was mistakenly re-added during merge, those assertions were moved to a few lines below: https://github.com/tigerbeetle/tigerbeetle/blob/65b913d5581ab8453484c4f514467c6db721452e/src/state_machine.zig#L3677

batiati committed:
s/code/status

batiati committed:
Fix typos

batiati committed:
s/{accounts,transfers}_results/{account,transfer}_results
